python-checks/src/main/java/org/sonar/python/checks/InfiniteRecursionCheck.java

```diff
@@ -0,0 +1,279 @@
+    Set<CfgBlock> pushedBlocks = new HashSet<>();
+    Deque<CfgBlock> blockToVisit = new ArrayDeque<>();
+    blockToVisit.addLast(cfg.start());
+    pushedBlocks.add(cfg.start());
+    while (!blockToVisit.isEmpty()) {
+      CfgBlock block = blockToVisit.removeFirst();
+      if (block == cfg.end()) {
+        return true;
+      }
+      List<Tree> blockRecursiveCalls = recursiveCallCollector.findRecursiveCalls(block.elements());
+      if (!blockRecursiveCalls.isEmpty()) {
+        allRecursiveCalls.addAll(blockRecursiveCalls);
+      } else {
+        block.successors().stream().filter(pushedBlocks::add).forEach(blockToVisit::addLast);
+      }
+    }
+    return recursiveCallCollector.functionSymbolHasBeenReassigned;
+  }
+
+  private static class RecursiveCallCollector extends BaseTreeVisitor {
+
+    private final boolean isMethod;
+    private final Symbol functionSymbol;
+    @Nullable
+    private final Symbol selfSymbol;
+    // Classes can not only be compared by their string names with the current semantic
+    @Nullable
+    private final String className;
+    private boolean functionSymbolHasBeenReassigned = false;
+    private final List<Tree> recursiveCalls = new ArrayList<>();
+
+    private RecursiveCallCollector(FunctionDef currentFunction, Symbol functionSymbol) {
+      isMethod = currentFunction.isMethodDefinition();
+      this.functionSymbol = functionSymbol;
+      if (isMethod) {
+        boolean isStatic = currentFunction.decorators().stream()
+          .map(Decorator::name).map(DottedName::names).map(d -> d.get(d.size() - 1)).map(Name::name)
+          .anyMatch(decorator -> decorator.equals("staticmethod") || decorator.equals("classmethod"));
+        if (isStatic) {
+          selfSymbol = null;
+          className = findParentClassName(currentFunction);
+        } else {
+          selfSymbol = findSelfParameterSymbol(currentFunction);
+          className = null;
+        }
+      } else {
+        selfSymbol = null;
+        className = null;
+      }
+    }
+
+    private List<Tree> findRecursiveCalls(List<Tree> elements) {
+      recursiveCalls.clear();
+      elements.forEach(element -> element.accept(this));
+      return recursiveCalls;
+    }
+
+    @Override
+    public void visitCallExpression(CallExpression callExpression) {
+      Expression callee = callExpression.callee();
+      if (matchesLookupFunction(callee) || matchesLookupMethod(callee)) {
+        recursiveCalls.add(callee);
+      }
+      super.visitCallExpression(callExpression);
+    }
+
+    @Override
+    public void visitFunctionDef(FunctionDef pyFunctionDefTree) {
+      // ignore
+    }
+
+    @Override
+    public void visitLambda(LambdaExpression pyLambdaExpressionTree) {
+      // ignore
+    }
+
+    @Override
+    public void visitForStatement(ForStatement pyForStatementTree) {
+      // ignore, unexpected "for" element in cfg blocks
```

Comment:
It's not very clear to me why we're excluding for statements. Is it for a limitation in the CFG? In that case, please create a ticket for it and document in the unit tests

Reply:
Now, I think it's useless. It was probably useful before I exclude `lambda` or `sub-function`, but it was not the right workaround. So I removed this `visitForStatement` and added more test related to `for`.